### PR TITLE
Wet bulb freezing level

### PR DESCRIPTION
**Pull request by mspelman07**
This PR adds in the ability to calculate a Wet bulb freezing level from a wet-bulb temperature on height levels.

Most of the changes for this PR are simply renaming variables to be more agnostic and apply to height and pressure coordinates rather than just pressure. The main changes are for the plugin ExtractPressureLevel (renamed to ExtractLevel). The core changes in this PR are:

- Plugin now looks first for a pressure coordinate and then a height coordinate. Depending on which it finds will determine whether a height or pressure level are extracted.
- In writing this plugin I found that stratify has an EXTRAPOLATE_NEAREST keyword which will automatically do what fill_in_bounds was previously doing so I have replaced this.
- I have also added an extra argument to the stratify interpolation of rising=False. This tells stratify which direction we expect the variable to change and solves a problem if there was a temperature inversion between the first two valid points (see notebook below)  
- CLI added to calculate wet bulb freezing level 

To show my changes to the ExtractLevel plugin haven't changed how the plugin works (except where they should change it) I have plotted a few profiles of what it extracted previously and what it extracts now (https://github.com/MetOffice/improver_aux/blob/extractpressure_behaviour/notebooks/Checking_changes_to_ExtractPressureLevels.ipynb)

Acceptance_test_data: https://github.com/metoppv/improver_test_data/pull/30

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Just a doc-string change needed.

A comment for your notebook: Surface temperature inversions like your final example are quite common, although they are usually much shallower than that. We won't have seen this in the hail output because surface temperature inversions are associated with stable, nocturnal, non-convective weather, when hail is not possible.

File: improver/cli/wet_bulb_freezing_level.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+from improver import cli
+
+
+@cli.clizefy
+@cli.with_output
+def process(wet_bulb_temperature: cli.inputcube):
+    """Module to generate wet-bulb freezing level.
+
+    The height level at which the wet-bulb temperature first drops below 237.15K
```
[the remark this answers is not in the document]
Comment (mspelman07) [reply improver/cli/wet_bulb_freezing_level.py]: Done

File: improver/cli/wet_bulb_freezing_level.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    (0 degrees Celsius) is extracted from the wet-bulb temperature cube.
+
+    In grid squares where the temperature never goes below 237.15K the highest
+    height level on the cube is returned. In grid square where the temperature
+    starts below 237.15K the lowest height on the cube is returned.
```
[the remark this answers is not in the document]
Comment (mspelman07) [reply improver/cli/wet_bulb_freezing_level.py]: Done

**Review by MoseleyS — APPROVED** [review]

**Review by brhooper — COMMENTED** [review]
Thanks @mspelman07, I've just added a few comments, mostly about docstrings.

File: improver/utilities/cube_extraction.py [hunk 3]
```diff
@@ -364,177 +368,139 @@ def fill_invalid(self, cube: Cube):
...
+        return template_cube
 
-    def process(self, variable_on_pressure_levels: Cube) -> Cube:
-        """Extracts the pressure level where the environment
-        variable first intersects self.value_of_pressure_level starting at a pressure value
-        near the surface and ascending in altitude from there.
-        Where the pressure surface falls outside the available data, the maximum or minimum
-        pressure will be returned, even if the source data has no value at that point.
+    def process(self, variable_on_levels: Cube) -> Cube:
+        """Extracts the pressure or height level (depending on which is present on the cube)
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/utilities/cube_extraction.py:455]: For Met Office StaGE data, there are no data that have both a pressure and a height coordinate, and I don't think it is possible to have one as the relationship between pressure and height requires knowledge of altitude and density. We have data of height of pressure levels for this purpose.
While it is possible to construct a cube with both a height and a pressure coordinate, it doesn't make sense to do so. Perhaps the safest approach is to raise a `NotImplementedError` if an input cube contains both these coordinates.

File: improver/cli/wet_bulb_freezing_level.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+@cli.clizefy
+@cli.with_output
+def process(wet_bulb_temperature: cli.inputcube):
+    """Module to generate wet-bulb freezing level.
+
+    The height level at which the wet-bulb temperature first drops below 273.15K
+    (0 degrees Celsius) is extracted from the wet-bulb temperature cube.
+
+    In grid squares where the temperature never goes below 273.15K the highest
+    height level on the cube is returned. In grid square where the temperature
```
[the remark this answers is not in the document]
Comment (mspelman07) [reply improver/cli/wet_bulb_freezing_level.py]: Done

File: improver/cli/wet_bulb_freezing_level.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+from improver import cli
+
+
+@cli.clizefy
+@cli.with_output
+def process(wet_bulb_temperature: cli.inputcube):
+    """Module to generate wet-bulb freezing level.
+
+    The height level at which the wet-bulb temperature first drops below 273.15K
```
[the remark this answers is not in the document]
Comment (mspelman07) [reply improver/cli/wet_bulb_freezing_level.py:42]: I've extended the sentence to clarify what I meant.

File: improver/utilities/cube_extraction.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mspelman07) [reply improver/utilities/cube_extraction.py:455]: I've taken Stephen's suggestions and added in an error in this case as well as an extra test to check this error.

[the remark this answers is not in the document]
Comment (brhooper) [reply improver/utilities/cube_extraction.py:455]: Stephen's suggestion is more sensible than mine, so that's probably for the best!

**Review by brhooper — APPROVED** [review]
Thanks @mspelman07, I'm happy to approve this now

**Pull request merged**